syntax = "proto3";

package instrosetta.interfaces.light_analysis.monochromator;

import "instrosetta/common/connection.proto";

service Monochromator {
    rpc Connect(ConnectRequest) returns (ConnectResponse) {}
    rpc Disconnect(DisconnectRequest) returns (DisconnectResponse) {}

    rpc GetWavelengthRange(GetWavelengthRangeRequest) returns (GetWavelengthRangeResponse) {}
    rpc GetWavelength(GetWavelengthRequest) returns (GetWavelengthResponse) {}
    rpc SetWavelength(SetWavelengthRequest) returns (SetWavelengthResponse) {}

    rpc GetGratingOptions(GetGratingOptionsRequest) returns (GetGratingOptionsResponse) {}
    rpc GetGrating(GetGratingRequest) returns (GetGratingResponse) {}
    rpc SetGrating(SetGratingRequest) returns (SetGratingResponse) {}

    rpc GetSlitWidthRange(GetSlitWidthRangeRequest) returns (GetSlitWidthRangeResponse) {}
    rpc GetSlitWidth(GetSlitWidthRequest) returns (GetSlitWidthResponse) {}
    rpc SetSlitWidth(SetSlitWidthRequest) returns (SetSlitWidthResponse) {}
}

message GetWavelengthRangeRequest {
    string name = 1; 
}

message GetWavelengthRangeResponse {
    string name = 1; 
    float minimum = 2;
    float maximum = 3;
    string units = 4;
}

message GetWavelengthRequest {
    string name = 1; 
}

message GetWavelengthResponse {
    string name = 1; 
    float magnitude = 2;
    string units = 3;
}

message SetWavelengthRequest {
    string name = 1; 
    float magnitude = 2;
    string units = 3;
}

message SetWavelengthResponse {
    string name = 1; 
    float magnitude = 2;
    string units = 3;
}


message GetGratingOptionsRequest {
    string name = 1; 
}

message GetGratingOptionsResponse {
    string name = 1; 
    repeated uint32 grating_ids = 2;
    
}

message GetGratingRequest {
    string name = 1; 
}

message GetGratingResponse {
    string name = 1; 
    uint32 grating_id = 2;
}

message SetGratingRequest {
    string name = 1; 
    uint32 grating_id = 2;
}

message SetGratingResponse {
    string name = 1; 
    uint32 grating_id = 2;
}


message GetSlitWidthRangeRequest {
    string name = 1; 
}

message GetSlitWidthRangeResponse {
    string name = 1; 
    float minimum = 2;
    float maximum = 3;
    string units = 4;
}

message GetSlitWidthRequest {
    string name = 1; 
}

message GetSlitWidthResponse {
    string name = 1; 
    float magnitude = 2;
    string units = 3;
}

message SetSlitWidthRequest {
    string name = 1; 
    float magnitude = 2;
    string units = 3;
}

message SetSlitWidthResponse {
    string name = 1; 
    float magnitude = 2;
    string units = 3;
}

